src/visibility_tracker.rs

```diff
@@ -606,198 +595,204 @@ fn register_name<'a>(
-        return;
-    }
+    let mut parents: HashMap<&'a Id, Vec<&'a Id>> = Default::default();
+    // Avoid cycles by keeping track of which items we're in the middle of visiting.
+    let mut currently_visited_items: HashSet<&'a str> = Default::default();
+
+    let mut stack = vec![Action::Process {
+        item,
+        parent_id: None,
+    }];
+
+    while let Some(action) = stack.pop() {
+        let (item, parent_id) = match action {
+            Action::Process { item, parent_id } => (item, parent_id),
+            Action::Remove { id } => {
+                // We are leaving this item. Remove it from the visited set.
+                let removed = currently_visited_items.remove(id.0.as_str());
+                assert!(removed);
+                continue;
+            }
+        };
 
-    let next_parent_id = Some(&item.id);
-    match &item.inner {
-        rustdoc_types::ItemEnum::Module(m) => {
-            for inner in m.items.iter().filter_map(|id| crate_.index.get(id)) {
-                visit_root_reachable_public_items(
-                    crate_,
-                    parents,
-                    traversal_state,
-                    currently_visited_items,
-                    inner,
-                    next_parent_id,
-                );
+        match item.visibility {
+            Visibility::Crate | Visibility::Restricted { .. } => {
+                // This item is not public, so we don't need to process it.
+                continue;
+            }
+            Visibility::Public => {} // Public item, keep going.
+            Visibility::Default => {
+                // Enum variants, and some impls and methods have default visibility:
+                // they are visible only if the type to which they belong is visible.
+                // However, we don't recurse into non-public items with this function, so
+                // reachable items with default visibility must be public.
             }
+        }
 
-            // Explicitly process items imported via globs inside this module,
-            // since the logic there is not item-wise: it requires
-            // knowledge of the other names defined in the module.
-            if let Some(glob_imports) = traversal_state.glob_imported_names_in_module.get(&item.id)
-            {
-                for inner_defn in glob_imports.values() {
-                    if let Some(inner_item) = crate_.index.get(inner_defn.current_id) {
+        let item_parents = parents.entry(&item.id).or_default();
+        if let Some(parent_id) = parent_id {
+            item_parents.push(parent_id);
+        }
+
+        if !currently_visited_items.insert(item.id.0.as_str()) {
+            // We found a cycle in the import graph, and we've already processed this item.
+            // Nothing more to do here.
+            continue;
+        }
+
+        let next_parent_id = Some(&item.id);
+        // queue up removing this from the visited list
```

Comment:
Just to match the style of the surrounding comments.
```suggestion
        // Queue up removing this from the visited list.
```

Reply:
Fixed (by removing the refactor)